File: eftc/eftc_app/report/top_paid_invoice/top_paid_invoice.py
# Copyright (c) 2023, NestorBird and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    columns, data = [], []
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    columns = [
        {
            "label": _("Customer"),
            "fieldname": "customer",
            "fieldtype": "Link",
            "options": "Customer",  
            "width": 200,
        },
        {
            "label": _("Sales Person"),
            "fieldname": "sales_person",
            "fieldtype": "Link",
            "options": "Sales Person",  
            "width": 200,
        },
        {
            "label": _("Amount"),
            "fieldname": "paid_amount",
            "fieldtype": "Float",
            "width": 200,
        }
    ]
    return columns

def get_data(filters):
    conditions = get_conditions(filters)
    main = frappe.db.sql(f"""
        SELECT
            si.name AS invoice,
            si.customer AS customer,
            st.sales_person AS sales_person,
            si.grand_total AS paid_amount
        FROM
            `tabSales Invoice` AS si
        JOIN
            `tabSales Team` AS st ON si.name = st.parent
        WHERE
            si.docstatus = 1
            AND si.status = 'Paid'
            {conditions}
        ORDER BY
            si.grand_total DESC
            LIMIT 5 
    """, as_dict=True)
    return main

def get_conditions(filters):
    conditions = []
    if filters.get("sales_person"):
        employee = frappe.db.get_value("Employee",{"user_id":filters['sales_person']},"name")
        sales_man = frappe.db.get_value("Sales Person",{"employee":employee},"name")
        if sales_man:
            conditions.append(f"AND st.sales_person = {frappe.db.escape(sales_man)}")
    if filters.get("company"):
        conditions.append(f"AND si.company = {frappe.db.escape(filters['company'])}")
    return " ".join(conditions)

File: eftc/eftc_app/report/top_paid_invoice/top_paid_invoice.js
// Copyright (c) 2023, NestorBird and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Top Paid Invoice"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",  // Change "Company" to the actual doctype of your company
			"reqd": 1,  // Make it mandatory if needed
			"default": frappe.defaults.get_user_default("Company")
		},
		
		{
			fieldname: "sales_person",
			label: __("Sales Person"),
			fieldtype: "Select",
			options: get_sales_person(),
			default:"All"
		},
	],
	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname.includes('variance')) {

			if (data[column.fieldname] < 0) {
				value = "<span style='color:red'>" + value + "</span>";
			}
			else if (data[column.fieldname] > 0) {
				value = "<span style='color:green'>" + value + "</span>";
			}
		}

		return value;
	}
};



function get_sales_person(){
	var sales_per_list = []
	frappe.call({
	    method:"eftc.eftc_app.report.sales_person_target_variance_based_on_item_group_for_chart.sales_person_target_variance_based_on_item_group_for_chart.get_sales_person",
	    async:true,
	    callback:function(r) {
	        if (r.message) {
	        	r.message.forEach(function(e){
	        		sales_per_list.push(e)
	        		
	        	})
	        }
	    }
	});
	return sales_per_list
}